torch/lib/THD2/FileStore.cpp

```diff
@@ -0,0 +1,270 @@
+            "Unable to open " + path);
+      }
+      break;
+    }
+  }
+
+  ~File() {
+    ::close(fd_);
+  }
+
+  Lock lockShared() {
+    return Lock(fd_, LOCK_SH);
+  }
+
+  Lock lockExclusive() {
+    return Lock(fd_, LOCK_EX);
+  }
+
+  off_t seek(off_t offset, int whence) {
+    auto rv = lseek(fd_, offset, whence);
+    SYSASSERT(rv);
+    return rv;
+  }
+
+  off_t tell() {
+    auto rv = lseek(fd_, 0, SEEK_CUR);
+    SYSASSERT(rv);
+    return rv;
+  }
+
+  off_t size() {
+    auto pos = tell();
+    auto size = seek(0, SEEK_END);
+    seek(pos, SEEK_SET);
+    return size;
+  }
+
+  void writeString(const std::string& str) {
+    uint32_t len = str.size();
+    SYSCHECK(write(fd_, &len, sizeof(len)));
+    SYSCHECK(write(fd_, str.c_str(), len));
+  }
+
+  void readString(std::string& str) {
+    uint32_t len;
+    SYSCHECK(read(fd_, &len, sizeof(len)));
+    std::vector<char> buf(len);
+    SYSCHECK(read(fd_, buf.data(), len));
+    str.assign(buf.begin(), buf.end());
+  }
+
+  void writeVector(const std::vector<char>& data) {
+    uint32_t len = data.size();
+    SYSCHECK(write(fd_, &len, sizeof(len)));
+    SYSCHECK(write(fd_, data.data(), len));
+  }
+
+  void readVector(std::vector<char>& data) {
+    uint32_t len;
+    SYSCHECK(read(fd_, &len, sizeof(len)));
+    data.resize(len);
+    SYSCHECK(read(fd_, data.data(), len));
+  }
+
+ protected:
+  int fd_;
+};
+
+} // namespace
+
+FileStore::FileStore(const std::string& path)
+    : Store(),
+      path_(path),
+      pos_(0) {
+}
+
+FileStore::~FileStore() {
+}
+
+void FileStore::set(const std::string& key, const std::vector<char>& value) {
+  File file(path_, O_RDWR | O_CREAT);
+  auto lock = file.lockExclusive();
+  file.seek(0, SEEK_END);
+  file.writeString(key);
+  file.writeVector(value);
+}
+
+std::vector<char> FileStore::get(const std::string& key) {
+  while (cache_.count(key) == 0) {
+    File file(path_, O_RDONLY);
+    auto lock = file.lockShared();
+    auto size = file.size();
+    if (size == pos_) {
+      // No new entries; sleep for a bit
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+      continue;
+    }
+
+    std::string tmpKey;
+    std::vector<char> tmpValue;
+    file.seek(pos_, SEEK_SET);
+    while (size > pos_) {
+      file.readString(tmpKey);
+      file.readVector(tmpValue);
+      cache_[tmpKey] = std::move(tmpValue);
+      pos_ = file.tell();
+    }
+  }
+
+  return cache_[key];
+}
+
+int64_t FileStore::add(const std::string& key, int64_t i) {
+  File file(path_, O_RDWR | O_CREAT);
+  auto lock = file.lockExclusive();
+  auto size = file.size();
+  if (size > pos_) {
+    // New entries; update cache since this key may have been updated
+    std::string tmpKey;
+    std::vector<char> tmpValue;
+    file.seek(pos_, SEEK_SET);
+    while (size > pos_) {
+      file.readString(tmpKey);
+      file.readVector(tmpValue);
+      cache_[tmpKey] = std::move(tmpValue);
+      pos_ = file.tell();
+    }
+  }
```

Comment:
Yep, I didn't do this at first because I didn't want to move the File and Lock classes into the header. But I can work around this by passing some refs around.